### ISSUE TITLE
attempt local face detection

### ISSUE THREAD

**Issue by afeld** [afeld/mustachio#38]
The app keeps hitting the Rekognition API quota, and doing detection in a library would make the responses happen _much_ faster. Suggested libraries (based on a couple hours of research):
- http://cmp.felk.cvut.cz/~uricamic/flandmark/
- ASM
  - https://code.google.com/p/asmlib-opencv/
  - http://blog.greatyao.me/2012/05/06/Introduction-to-ASMLibrary/
- LBP
- http://webcache.googleusercontent.com/search?q=cache:ioIRRRgx0xYJ:answers.opencv.org/question/313/facial-feature-detection/+&cd=2&hl=en&ct=clnk&gl=us
- [OpenBR](http://openbiometrics.org/)
- [facedetect](https://www.thregr.org/~wavexx/software/facedetect/)
- [OpenFace](https://github.com/TadasBaltrusaitis/OpenFace)
- [clmtrackr](https://github.com/auduno/clmtrackr)

_will keep adding to this list_

**Comment by afeld** [comment]
Using Docker might make this a lot easier:

https://github.com/steeve/docker-opencv

**Comment by afeld** [comment]
* [OpenBR](http://openbiometrics.org/)

**Comment by netham91** [comment]
Did you give tracking.js a try? https://trackingjs.com/examples/face_tag_friends.html

**Comment by afeld** [comment]
I haven't! Would happily accept a pull request for anything. _[Edit: Previous comment was deleted...?]_

**Comment by afeld** [comment]
Extra points for OpenCV:

http://www.pyimagesearch.com/2016/06/20/detecting-cats-in-images-with-opencv/

**Comment by mathewmeconry** [comment]
maybe you can use this:
https://trackingjs.com/

I know it not ruby but maybe reconding in nodejs would lead to a faster application

**Comment by afeld** [comment]
Most of the actual detection code for the libraries above is in C/C++, but I'm open to anything at this point!

Note that any new detection code would be added to https://github.com/afeld/face_detect.

**Comment by neufeldtech** [comment]
@afeld I've been working on a small node.js face detection app and have a docker container built if you think it might be of use. https://github.com/neufeldtech/funnyface.

**Comment by ChrisCinelli** [comment]
https://www.npmjs.com/package/faced

**Comment by darkyen** [comment]
Faced looks really promising and since its using OpenCV this should be incredibly fast :D

**Comment by melv-n** [comment]
Faced looks good!

**Comment by denimamab** [comment]
Hey !
Checkout this API : [FacePlusPlus](https://www.faceplusplus.com/face-detection/)

**Comment by kamlekar** [comment]
Hey @afeld check above suggestions. They look promising.

**Comment by afeld** [comment]
Yep, open to any of them, with a strong preference for local rather than an external API (for $$ and latency reasons). Pull requests welcome!

**Comment by Cbeck527** [comment]
I've been itching to whip something up recently- AWS Lambda has come a long way w/r/t python packaging. I recently learned you can boot a aws linux docker container and build out your compiled deps, in this case `opencv`, and bundle that compiled virtualenv with your deployment payload. 

Could be neat! I hope to have some free time to try this in the near future. Our hipchat has had a severe mustache shortage :flushed:
